src/batch.rs

```diff
@@ -0,0 +1,297 @@
+
+// Shim to generate a random 128bit value in a [u64; 4], without
+// importing `rand`.
+fn gen_128_bits<R: RngCore + CryptoRng>(mut rng: R) -> [u64; 4] {
+    let mut bytes = [0u64; 4];
+    bytes[0] = rng.next_u64();
+    bytes[1] = rng.next_u64();
+    bytes
+}
+
+enum Inner {
+    SpendAuth {
+        vk_bytes: VerificationKeyBytes<SpendAuth>,
+        sig: Signature<SpendAuth>,
+        c: Scalar,
+    },
+    Binding {
+        vk_bytes: VerificationKeyBytes<Binding>,
+        sig: Signature<Binding>,
+        c: Scalar,
+    },
+}
+
+impl
+    From<(
+        VerificationKeyBytes<SpendAuth>,
+        Signature<SpendAuth>,
+        Scalar,
+    )> for Inner
+{
+    fn from(
+        tup: (
+            VerificationKeyBytes<SpendAuth>,
+            Signature<SpendAuth>,
+            Scalar,
+        ),
+    ) -> Self {
+        let (vk_bytes, sig, c) = tup;
+        Inner::SpendAuth { vk_bytes, sig, c }
+    }
+}
+
+impl From<(VerificationKeyBytes<Binding>, Signature<Binding>, Scalar)> for Inner {
+    fn from(tup: (VerificationKeyBytes<Binding>, Signature<Binding>, Scalar)) -> Self {
+        let (vk_bytes, sig, c) = tup;
+        Inner::Binding { vk_bytes, sig, c }
+    }
+}
+
+/// A batch verification item.
+///
+/// This struct exists to allow batch processing to be decoupled from the
+/// lifetime of the message. This is useful when using the batch verification API
+/// in an async context.
+pub struct Item {
+    inner: Inner,
+}
+
+impl<'msg, M: AsRef<[u8]>>
+    From<(
+        VerificationKeyBytes<SpendAuth>,
+        Signature<SpendAuth>,
+        &'msg M,
+    )> for Item
+{
+    fn from(
+        tup: (
+            VerificationKeyBytes<SpendAuth>,
+            Signature<SpendAuth>,
+            &'msg M,
+        ),
+    ) -> Self {
+        let (vk_bytes, sig, msg) = tup;
+        // Compute c now to avoid dependency on the msg lifetime.
+        let c = HStar::default()
+            .update(&sig.r_bytes[..])
+            .update(&vk_bytes.bytes[..])
+            .update(msg)
+            .finalize();
+        Self {
+            inner: Inner::SpendAuth { vk_bytes, sig, c },
+        }
+    }
+}
+
+impl<'msg, M: AsRef<[u8]>> From<(VerificationKeyBytes<Binding>, Signature<Binding>, &'msg M)>
+    for Item
+{
+    fn from(tup: (VerificationKeyBytes<Binding>, Signature<Binding>, &'msg M)) -> Self {
+        let (vk_bytes, sig, msg) = tup;
+        // Compute c now to avoid dependency on the msg lifetime.
+        let c = HStar::default()
+            .update(&sig.r_bytes[..])
+            .update(&vk_bytes.bytes[..])
+            .update(msg)
+            .finalize();
+        Self {
+            inner: Inner::Binding { vk_bytes, sig, c },
+        }
+    }
+}
+
+// This would ideally be an associated type with `Verifier` but
+// generic associated types are unstable:
+// https://github.com/rust-lang/rust/issues/44265
+type BatchTuple<T: SigType> = (VerificationKeyBytes<T>, Signature<T>, Scalar);
```

Comment:
This is leftover from my submethod that I didn't commit because I was running into `Borrow` issues with it.

Reply:
OK, did you end up with wanting to use the submethod or wanting to remove the type?

Reply:
Removing based on feedback 👍